File: mini_ep1/mini_ep1.py
# Mini EP1 - Linguagem lenta
# Nome: Lara Ayumi Nagamatsu
# NUSP: 9910568
# como rodar: 
#   digite no terminal:
#   1.1 python3 mini_ep1.py
#   obs: caso queira medir o tempo, rode no terminal:
#   1.2 time python3 mini_ep1.py


def main():
    N = 1 << int(input("Digite o N: "))
    primos = 0
    primosEspeciais = 0
    crivo = [None] * (N+1)

    for i in range(2, N):
        crivo[i] = True

    for i in range(2, N):
        if crivo[i] == True:
            primos = primos + 1
            
            if i%4 != 3:
                primosEspeciais = primosEspeciais +1
            
            for j in range(i*2, N, i):
                crivo[j] = False

    print("Primos: ", primos, "\nPrimos especiais: ", primosEspeciais)


main()

File: mini_ep1/mini_ep1.c
/* Mini EP1 - Linguagem Rápida
 * Nome: Lara Ayumi Nagamatsu
 * NUSP: 9910568
 * como compilar:  
 *   digite no terminal:
 *   1.  make mini_rapido
 *   isso gerará o executável "mini", rode no terminal:
 *   2.1  ./mini
 *   obs: caso queira testar o tempo, rode no terminal:
 *   2.2  time ./mini
 */

#include <stdio.h>  //printf, scanf
#include <stdlib.h> //malloc, free

int main() {
    int N;
    printf("Insira o N: ");
    scanf("%d", &N);
    N = 1 << N;       // operação de bitshift

    int primos = 0;
    int primosEspeciais = 0;
    int *crivo = malloc((N+1)*sizeof(int)); // cria lista de booleanos com n+1 elementos

    for(int i = 2; i <= N; i++) {
        crivo[i] = 1;
    }

    for(int i = 2; i <= N; i++) {
        if(crivo[i] == 1) {
            //printf("%d ", i);
            primos++;
            
            if(i%4 != 3)
                primosEspeciais++;
            
            for(int j = i*2; j <= N; j = j+i)
                crivo[j] = 0;
        }
    }

    printf("Primos: %d\nPrimos especiais: %d\n", primos, primosEspeciais);
    
    free(crivo);  // libera memória alocada

    return 0;
}